cubeviz/data_factories/__init__.py

```diff
@@ -2,3 +2,287 @@
+            cfg = yaml.load(ymlfile)
+
+            self._name = cfg['name']
+            self._type = cfg['type']
+
+            try:
+                self._priority = int(cfg.get('priority', 0))
+            except:
+                self._priority = 0
+
+            self._configuration = cfg['match']
+
+            self._data = cfg.get('data', None)
+
+
+    def load_data(self, data_filename):
+        """
+        Load the data based on the extensions defined in the matching YAML file.  THen
+        create the datacube and return it.
+
+        :param data_filename:
+        :return:
+        """
+        hdulist = fits.open(data_filename)
+
+        try:
+            flux_index = int(self._data['FLUX'])
+        except:
+            flux_index = self._data['FLUX']
+
+        flux = hdulist[flux_index]
+        flux_data = flux.data
+
+        try:
+            error_index = int(self._data['ERROR'])
+        except:
+            error_index = self._data['ERROR']
+        var_data = hdulist[error_index].data
+
+        if not self._data['DQ'] == 'None':
+            mask_data = hdulist[self._data['DQ']].data
+        else:
+            mask_data = np.empty(flux_data.shape)
+
+        label = "{}: {}".format(self._name, splitext(basename(data_filename))[0])
+
+        data = Data(label=label)
+
+        data.coords = coordinates_from_header(flux.header)
+        data.meta[CUBEVIZ_LAYOUT] = self._name
+
+        data.add_component(component=flux_data, label=FLUX)
+        data.add_component(component=var_data, label=ERROR)
+        data.add_component(component=mask_data, label=MASK)
+
+        return data
+
+    def matches(self, filename):
+        """
+        Main call to which we pass in the file to see if it matches based
+        on the criteria in the config file.
+
+        :param filename:
+        :return:
+        """
+        self._fits = fits.open(filename)
+
+        logger.debug('Checking {} to see if matches {}'.format(filename, self._config_file))
+
+
+        # Now call the internal processing.
+        matches = self._process('all', self._configuration['all'])
+
+        if matches:
+            logger.debug('{} matches {}'.format(self._config_file, filename))
+            return True
+        else:
+            return False
+
+    def _process(self, key, conditional):
+        """
+        Internal processing. This will get called numerous times recursively.
+
+        :param key: The type of check we want to do
+        :param conditional: The thing we are checking
+        :return:
+        """
+
+        if 'all' == key:
+            return self._all(conditional)
+        elif 'any' == key:
+            return self._any(conditional)
+        elif 'equal' == key:
+            return self._equal(conditional)
+        elif 'startswith' == key:
+            return self._startswith(conditional)
+        elif 'extension_name' == key:
+            return self._exists(conditional)
```

Comment:
Poor Python without a `switch` statement. Sometimes I wonder if this kind of code would be improved by a `dict`, but it's not really important.